include/hipSYCL/sycl/libkernel/sscp/builtins.hpp

```diff
@@ -459,6 +459,22 @@ HIPSYCL_BUILTIN T __hipsycl_mul24(T x, T y) noexcept {
   return __hipsycl_sscp_mul24_u32(x, y);
 }
 
+template <class T,
+          std::enable_if_t<
+              (std::is_integral_v<T> && sizeof(T) <= 4),
+              int> = 0>
+HIPSYCL_BUILTIN T __hipsycl_popcount(T x) noexcept {
+  return __hipsycl_sscp_popcount_u32(static_cast<__hipsycl_uint32>(x));
```

Comment:
You also need to add definitions for these functions for the various backends (see `src/libkernel`). Otherwise the SSCP compilation path will fail (`--hipsycl-targets=generic`). Since SSCP can target all supported device backends, you can also test this with the hardware you have available.

Reply:
I have tried `generic`, however, it keeps producing this error:
```
[hipSYCL Warning] OpenCL device NVIDIA GeForce RTX 2080 with Max-Q Design does not have a valid USM provider. Memory allocations are not possible on that device.
[hipSYCL Warning] OpenCL device Intel(R) FPGA Emulation Device does not support Intel USM extensions; falling back to fine-grained system SVM. USM pointer info queries have limited support.
NVIDIA GeForce RTX 2080 with Max-Q Design
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code object creation has failed
```

Reply:
Actually, even an empty kernel results in error too whenever I used `export OPENSYCL_TARGETS=generic`:
```cpp
#include <sycl/sycl.h>
int main() {
    sycl::queue q;
    q.submit([&] (sycl::handler &cgh) {
        cgh.single_task<class single>([=] {
        });
    });
}
```
Is this an expected behaviour? I'm using RTX 2080 and CUDA 12.2.

Reply:
Are you building with `-g`? There is currently an issue with that in the CUDA SSCP path, as it seems the CUDA tools really don't like debug information. See issue #1062
The solution for now is to not enable debug information.

Reply:
No, I didn't use `-g`.
```
(opensycl) ➜  sycl2020test git:(master) ✗ syclcc -O3 main.cpp
(opensycl) ➜  sycl2020test git:(master) ✗ ./a.out     
[hipSYCL Warning] OpenCL device NVIDIA GeForce RTX 2080 with Max-Q Design does not have a valid USM provider. Memory allocations are not possible on that device.
[hipSYCL Warning] OpenCL device Intel(R) FPGA Emulation Device does not support Intel USM extensions; falling back to fine-grained system SVM. USM pointer info queries have limited support.
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object construction failed
(opensycl) ➜  sycl2020test git:(master) ✗ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Tue_Jun_13_19:16:58_PDT_2023
Cuda compilation tools, release 12.2, V12.2.91
Build cuda_12.2.r12.2/compiler.32965470_0
(opensycl) ➜  sycl2020test git:(master) ✗ 
```

Reply:
It seems that this function is returning CUDA_ERROR_INVALID_PTX:
```
  auto err = cuModuleLoadDataEx(
      &module, static_cast<void *>(const_cast<char *>(source.c_str())),
      0, nullptr, nullptr);
```
The source is:
```
(gdb) p -elements unlimited -- source
$8 = "//\n// Generated by LLVM NVPTX Back-End\n//\n\n.version 7.5\n.target sm_75\n.address_size 64\n\n", ' ' <repeats 40 times>, "// Start of file scope inline assembly\n.globl _ZSt21ios_base_library_initv\n\n", ' ' <repeats 40 times>, "// End of file scope inline assembly\n\t// .globl\t_Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch11single_taskIZZ4mainENK3$_0clERNS0_4sycl7handlerEEUlvE_EEEvRKT_ // -- Begin function _Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch11single_taskIZZ4mainENK3$_0clERNS0_4sycl7handlerEEUlvE_EEEvRKT_\n", ' ' <repeats 40 times>, "// @\"_Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch11single_taskIZZ4mainENK3$_0clERNS0_4sycl7handlerEEUlvE_EEEvRKT_\"\n.visible .entry _Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch11single_taskIZZ4mainENK3$_0clERNS0_4sycl7handlerEEUlvE_EEEvRKT_(\n\t.param .u8 _Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch11single_taskIZZ4mainENK3$_0clERNS0_4sycl7handlerEEUlvE_EEEvRKT__param_0\n)\n{\n\n\n// %bb.0:\n\tret;\n", ' ' <repeats 40 times>, "// -- End function\n}\n"
```
Hope that helps!

Reply:
* Which LLVM version are you using? We do not support development versions of LLVM.
* It's unlikely that this is the reason, but try setting the environment variable `HIPSYCL_VISIBILITY_MASK="omp;cuda" which prevents the OpenCL backend from loading in case there interference between the NVIDIA OpenCL drivers and CUDA.
* Does it happen only in your branch but also on `develop`? Undefined device functions such as the missing popcount implementation can be a cause for JIT failure.
* Please provide the IR that fails JIT. Set the environment variable `HIPSYCL_HCF_DUMP_DIRECTORY` to some existing directory; the runtime will then fill that directory with the HCF files contained in the application. HCF files store device code IR. You can either extract that IR yourself using `opensycl-hcf-tool` or provide the HCF files themselves.

Reply:
> Which LLVM version are you using? We do not support development versions of LLVM.

LLVM 15.0.7

> Does something change if you compile with -O3?

I have tried to compile the simple example with -O3 and without -O3, but they result in same error.

> It's unlikely that this is the reason, but try setting the environment variable HIPSYCL_VISIBILITY_MASK="omp;cuda" which prevents the OpenCL backend from loading in case there interference between the NVIDIA OpenCL drivers and CUDA.

Setting HIPSYCL_VISIBILITY_MASK="omp;cuda" result in same error. HIPSYCL_VISIBILITY_MASK="omp" worked, however.

> Does it happen only in your branch but also on develop? Undefined device functions such as the missing popcount implementation can be a cause for JIT failure.

Tried the latest `develop` (12f8c24d), still the same error.

> Add q.wait() add the end of your simple example. Otherwise it can happen that the application exits while submission/JIT is in progress, and you end up in a world of undefined behavior.

Sorry, I did't know that it's a standard to call q.wait() before queue destruction. Here's the error message after adding q.wait():

```
[hipSYCL Warning] OpenCL device NVIDIA GeForce RTX 2080 with Max-Q Design does not have a valid USM provider. Memory allocations are not possible on that device.
[hipSYCL Warning] OpenCL device Intel(R) FPGA Emulation Device does not support Intel USM extensions; falling back to fine-grained system SVM. USM pointer info queries have limited support.
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object construction failed
============== hipSYCL error report ============== 
hipSYCL has caught the following undhandled asynchronous errors: 

   0. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could not load module (error code = CU:218)
   1. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code object creation has failed
   2. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object construction failed
The application will now be terminated.
terminate called without an active exception
[1]    6034 IOT instruction (core dumped)  ./a.out
```

> Please provide the IR that fails JIT. Set the environment variable HIPSYCL_HCF_DUMP_DIRECTORY to some existing directory; the runtime will then fill that directory with the HCF files contained in the application. HCF files store device code IR. You can either extract that IR yourself using opensycl-hcf-tool or provide the HCF files themselves.

Running `opensycl-hcf-tool /tmp/hipsycl_object_4830165983659051074.hcf -x root` result in `Specified node does not have binary data attached.` error, not sure how to use this tool correctly. So I'll attach my HCF file here: https://github.com/archibate/taichi_public_files/blob/master/hipsycl_object_4830165983659051074.hcf

Reply:
Thanks, you need to give the HCF tool the internal path to the binary node that you want to extract, e.g.:
`opensycl-hcf-tool file.hcf -x root images llvm-ir.global > out.bc`

You can inspect the file structure using `opensycl-hcf-tool file.hcf -p root`.

Extracting and disassembling the IR in your HCF, I get:
```
; ModuleID = 'out.bc'
source_filename = "hello.cpp"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-linux-gnu"

module asm ".globl _ZSt21ios_base_library_initv"

; Function Attrs: mustprogress nofree norecurse nosync nounwind readnone willreturn
define void @"_Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch11single_taskIZZ4mainENK3$_0clERNS0_4sycl7handlerEEUlvE_EEEvRKT_"(i8 %0) local_unnamed_addr #0 {
  ret void
}

attributes #0 = { mustprogress nofree norecurse nosync nounwind readnone willreturn }

!llvm.module.flags = !{!0, !1, !2, !3}
!llvm.ident = !{!4}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 7, !"PIC Level", i32 2}
!2 = !{i32 7, !"PIE Level", i32 2}
!3 = !{i32 7, !"uwtable", i32 2}
!4 = !{!"clang version 15.0.7"}

```

It contains module-level inline assembly (`module asm`) which is probably causing the issue, and I have not seen before happening. It seems to be related to `std::ios`, so maybe your standard library might be using inline assembly globals that the SSCP compiler then cannot know that they only related to host code. Which standard library are you using, and in which version?

Reply:
I'm using libstdc++ 6.0.32.

Tried to compile my sample file with `-stdlib=libc++` option, however, result in linker error:
```
/usr/bin/ld: /tmp/hello-5d0d5c.o: in function `main':
hello.cpp:(.text+0x2c1): undefined reference to `hipsycl::rt::execution_hints::overwrite_with(std::__1::shared_ptr<hipsycl::rt::execution_hint>)'
/usr/bin/ld: hello.cpp:(.text+0x3bf): undefined reference to `hipsycl::rt::execution_hints::overwrite_with(std::__1::shared_ptr<hipsycl::rt::execution_hint>)'
/usr/bin/ld: hello.cpp:(.text+0x4ae): undefined reference to `hipsycl::rt::execution_hints::overwrite_with(std::__1::shared_ptr<hipsycl::rt::execution_hint>)'
/usr/bin/ld: hello.cpp:(.text+0xa8e): undefined reference to `hipsycl::rt::requirements_list::add_node_requirement(std::__1::shared_ptr<hipsycl::rt::dag_node>)'
/usr/bin/ld: hello.cpp:(.text+0xf81): undefined reference to `hipsycl::rt::dag_builder::add_kernel(std::__1::unique_ptr<hipsycl::rt::operation, std::__1::default_delete<hipsycl::rt::operation> >, hipsycl::rt::requirements_list const&, hipsycl::rt::execution_hints const&)'
/usr/bin/ld: /tmp/hello-5d0d5c.o: in function `hipsycl::sycl::queue::queue(hipsycl::sycl::context const&, std::__1::vector<hipsycl::sycl::device, std::__1::allocator<hipsycl::sycl::device> > const&, std::__1::function<void (std::__1::vector<std::exception_ptr, std::__1::allocator<std::exception_ptr> >)> const&, hipsycl::sycl::property_list const&)':
hello.cpp:(.text._ZN7hipsycl4sycl5queueC2ERKNS0_7contextERKNSt3__16vectorINS0_6deviceENS5_9allocatorIS7_EEEERKNS5_8functionIFvNS6_ISt13exception_ptrNS8_ISE_EEEEEEERKNS0_13property_listE[_ZN7hipsycl4sycl5queueC2ERKNS0_7contextERKNSt3__16vectorINS0_6deviceENS5_9allocatorIS7_EEEERKNS5_8functionIFvNS6_ISt13exception_ptrNS8_ISE_EEEEEEERKNS0_13property_listE]+0x17d): undefined reference to `hipsycl::rt::execution_hints::add_hint(std::__1::shared_ptr<hipsycl::rt::execution_hint>)'
/usr/bin/ld: hello.cpp:(.text._ZN7hipsycl4sycl5queueC2ERKNS0_7contextERKNSt3__16vectorINS0_6deviceENS5_9allocatorIS7_EEEERKNS5_8functionIFvNS6_ISt13exception_ptrNS8_ISE_EEEEEEERKNS0_13property_listE[_ZN7hipsycl4sycl5queueC2ERKNS0_7contextERKNSt3__16vectorINS0_6deviceENS5_9allocatorIS7_EEEERKNS5_8functionIFvNS6_ISt13exception_ptrNS8_ISE_EEEEEEERKNS0_13property_listE]+0x3e9): undefined reference to `hipsycl::rt::execution_hints::add_hint(std::__1::shared_ptr<hipsycl::rt::execution_hint>)'
/usr/bin/ld: /tmp/hello-5d0d5c.o: in function `hipsycl::sycl::queue::init()':
hello.cpp:(.text._ZN7hipsycl4sycl5queue4initEv[_ZN7hipsycl4sycl5queue4initEv]+0x16b): undefined reference to `hipsycl::rt::execution_hints::add_hint(std::__1::shared_ptr<hipsycl::rt::execution_hint>)'
/usr/bin/ld: hello.cpp:(.text._ZN7hipsycl4sycl5queue4initEv[_ZN7hipsycl4sycl5queue4initEv]+0x228): undefined reference to `hipsycl::rt::execution_hints::add_hint(std::__1::shared_ptr<hipsycl::rt::execution_hint>)'
/usr/bin/ld: hello.cpp:(.text._ZN7hipsycl4sycl5queue4initEv[_ZN7hipsycl4sycl5queue4initEv]+0x2bb): undefined reference to `hipsycl::rt::execution_hints::add_hint(std::__1::shared_ptr<hipsycl::rt::execution_hint>)'
/usr/bin/ld: /tmp/hello-5d0d5c.o:hello.cpp:(.text._ZN7hipsycl4sycl5queue4initEv[_ZN7hipsycl4sycl5queue4initEv]+0x345): more undefined references to `hipsycl::rt::execution_hints::add_hint(std::__1::shared_ptr<hipsycl::rt::execution_hint>)' follow
/usr/bin/ld: /tmp/hello-5d0d5c.o: in function `hipsycl::glue::throw_result(hipsycl::rt::result const&)':
hello.cpp:(.text._ZN7hipsycl4glue12throw_resultERKNS_2rt6resultE[_ZN7hipsycl4glue12throw_resultERKNS_2rt6resultE]+0x1a9): undefined reference to `hipsycl::rt::result::what() const'
/usr/bin/ld: /tmp/hello-5d0d5c.o: in function `hipsycl::sycl::unimplemented::exception(hipsycl::rt::result const&)':
hello.cpp:(.text._ZN7hipsycl4sycl13unimplementedCI2NS0_9exceptionEERKNS_2rt6resultE[_ZN7hipsycl4sycl13unimplementedCI2NS0_9exceptionEERKNS_2rt6resultE]+0x28): undefined reference to `hipsycl::rt::result::what() const'
/usr/bin/ld: /tmp/hello-5d0d5c.o: in function `hipsycl::sycl::runtime_error::exception(hipsycl::rt::result const&)':
hello.cpp:(.text._ZN7hipsycl4sycl13runtime_errorCI2NS0_9exceptionEERKNS_2rt6resultE[_ZN7hipsycl4sycl13runtime_errorCI2NS0_9exceptionEERKNS_2rt6resultE]+0x28): undefined reference to `hipsycl::rt::result::what() const'
/usr/bin/ld: /tmp/hello-5d0d5c.o: in function `hipsycl::sycl::kernel_error::exception(hipsycl::rt::result const&)':
hello.cpp:(.text._ZN7hipsycl4sycl12kernel_errorCI2NS0_9exceptionEERKNS_2rt6resultE[_ZN7hipsycl4sycl12kernel_errorCI2NS0_9exceptionEERKNS_2rt6resultE]+0x28): undefined reference to `hipsycl::rt::result::what() const'
/usr/bin/ld: /tmp/hello-5d0d5c.o: in function `hipsycl::sycl::accessor_error::exception(hipsycl::rt::result const&)':
hello.cpp:(.text._ZN7hipsycl4sycl14accessor_errorCI2NS0_9exceptionEERKNS_2rt6resultE[_ZN7hipsycl4sycl14accessor_errorCI2NS0_9exceptionEERKNS_2rt6resultE]+0x28): undefined reference to `hipsycl::rt::result::what() const'
/usr/bin/ld: /tmp/hello-5d0d5c.o:hello.cpp:(.text._ZN7hipsycl4sycl14nd_range_errorCI2NS0_9exceptionEERKNS_2rt6resultE[_ZN7hipsycl4sycl14nd_range_errorCI2NS0_9exceptionEERKNS_2rt6resultE]+0x28): more undefined references to `hipsycl::rt::result::what() const' follow
/usr/bin/ld: /tmp/hello-5d0d5c.o: in function `std::__1::unique_ptr<hipsycl::rt::operation, std::__1::default_delete<hipsycl::rt::operation> > hipsycl::rt::make_operation<hipsycl::rt::kernel_operation, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::vector<std::__1::unique_ptr<hipsycl::rt::backend_kernel_launcher, std::__1::default_delete<hipsycl::rt::backend_kernel_launcher> >, std::__1::allocator<std::__1::unique_ptr<hipsycl::rt::backend_kernel_launcher, std::__1::default_delete<hipsycl::rt::backend_kernel_launcher> > > >, hipsycl::rt::requirements_list>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::vector<std::__1::unique_ptr<hipsycl::rt::backend_kernel_launcher, std::__1::default_delete<hipsycl::rt::backend_kernel_launcher> >, std::__1::allocator<std::__1::unique_ptr<hipsycl::rt::backend_kernel_launcher, std::__1::default_delete<hipsycl::rt::backend_kernel_launcher> > > >, hipsycl::rt::requirements_list)':
hello.cpp:(.text._ZN7hipsycl2rt14make_operationINS0_16kernel_operationEJNSt3__112basic_stringIcNS3_11char_traitsIcEENS3_9allocatorIcEEEENS3_6vectorINS3_10unique_ptrINS0_23backend_kernel_launcherENS3_14default_deleteISC_EEEENS7_ISF_EEEENS0_17requirements_listEEEENSB_INS0_9operationENSD_ISJ_EEEEDpT0_[_ZN7hipsycl2rt14make_operationINS0_16kernel_operationEJNSt3__112basic_stringIcNS3_11char_traitsIcEENS3_9allocatorIcEEEENS3_6vectorINS3_10unique_ptrINS0_23backend_kernel_launcherENS3_14default_deleteISC_EEEENS7_ISF_EEEENS0_17requirements_listEEEENSB_INS0_9operationENSD_ISJ_EEEEDpT0_]+0x5f): undefined reference to `hipsycl::rt::kernel_operation::kernel_operation(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<std::__1::unique_ptr<hipsycl::rt::backend_kernel_launcher, std::__1::default_delete<hipsycl::rt::backend_kernel_launcher> >, std::__1::allocator<std::__1::unique_ptr<hipsycl::rt::backend_kernel_launcher, std::__1::default_delete<hipsycl::rt::backend_kernel_launcher> > > >, hipsycl::rt::requirements_list const&)'
clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
```
Seems src/runtime need to have same ABI as the compiled sample? I tried to compile OpenSYCL with libc++, but failed with similar linker error.

Reply:
I find that this symbol is introduced in `/usr/include/c++/13.1.1/iostream`:
```cpp
  // For construction of filebuffers for cout, cin, cerr, clog et. al.
  // When the init_priority attribute is usable, we do this initialization
  // in the compiled library instead (src/c++98/globals_io.cc).
#if !__has_attribute(__init_priority__)
  static ios_base::Init __ioinit;
#elif defined(_GLIBCXX_SYMVER_GNU)
  __extension__ __asm (".globl _ZSt21ios_base_library_initv");
#endif
```
Maybe we should undefine _GLIBCXX_SYMVER_GNU here as LLVM doesn't recognize `.globl` instruction?

Reply:
After modifying `iostream` manually to:
```cpp
#elif defined(_GLIBCXX_SYMVER_GNU) && !defined(__OPENSYCL__)
```
The empty kernel sample finally worked, is it possible to walk around without requiring user to manually mock system headers?

Reply:
> Seems src/runtime need to have same ABI as the compiled sample? 

Yes, the headers as used by the application in general exchange C++ objects with the runtime library, so both need to be ABI compatible. Further requirements could be imposed by dependencies, such as boost.

> I find that this symbol is introduced in /usr/include/c++/13.1.1/iostream

Good to know, so it was introduced in libstdc++ 13 headers, which explains why I haven't seen it so far.

> The empty kernel sample finally worked, is it possible to walk around without requiring user to manually mock system headers?

Thanks for investigating and confirming the `module asm()` part is indeed the root cause of the issue.
I think we should modify the SSCP compiler to strip module-level inline assembly from device IR. I cannot see any case where module-level inline assembly might be needed in device code (or can even be supported in device code), so it should be safe to just remove it if we find it.

Reply:
Can you please try if #1123 fixes the issue for you?

Reply:
> Can you please try if #1123 fixes the issue for you?

Thank you! It works for my empty kernel sample!
However, when trying to run another complicated sample:
```cpp
#include <sycl/sycl.hpp>

class KT_exclusive_scan;
class KT_exclusive_scan_paste;

inline void exclusive_scan(sycl::queue &q, sycl::buffer<unsigned> &hist_group) {
    if (hist_group.size() <= 1) return;
    sycl::buffer<unsigned> glob_sum{(hist_group.size() + 255) / 256};
    q.submit([&] (sycl::handler &cgh) {
        sycl::accessor hist{hist_group, cgh, sycl::read_write};
        sycl::accessor gsum{glob_sum, cgh, sycl::write_only, sycl::no_init};
        cgh.parallel_for<KT_exclusive_scan>(sycl::nd_range<1>{glob_sum.size() * 256, 256}, [=] (sycl::nd_item<1> it) {
            int ii = it.get_local_id(0);
            int gi = it.get_group(0);
            int i = it.get_global_id(0);
            unsigned val = sycl::inclusive_scan_over_group(it.get_group(), i < hist.size() ? hist[i] : 0u, std::plus<>{});
            if (ii == 255) {
                gsum[gi] = val;
            } else if (i + 1 < hist.size()) {
                hist[i + 1] = val;
            }
            if (ii == 0) {
                hist[i] = 0;
            }
        });
    });
    if (glob_sum.size() > 1) {
        exclusive_scan(q, glob_sum);
        q.submit([&] (sycl::handler &cgh) {
            sycl::accessor hist{hist_group, cgh, sycl::read_write};
            sycl::accessor gsum{glob_sum, cgh, sycl::read_only};
            cgh.parallel_for<KT_exclusive_scan_paste>(sycl::nd_range<1>{glob_sum.size() * 256, 256}, [=] (sycl::nd_item<1> it) {
                int gi = it.get_group(0);
                int i = it.get_global_id(0);
                if (i < hist.size())
                    hist[i] += gsum[gi];
            });
        });
    }
}

class KT_radix_sort_histogram;
class KT_radix_sort_scatter;

template <
    sycl::memory_order memord = sycl::memory_order_relaxed,
    sycl::memory_scope memscp = sycl::memory_scope_work_group,
    class T>
sycl::atomic_ref<T, memord, memscp> atomic_ref(T &t) {
    return sycl::atomic_ref<T, memord, memscp>{t};
}

inline void radix_sort(sycl::queue &q, sycl::buffer<unsigned> &buf) {
    sycl::buffer<unsigned> buf_next{buf.size()};
    sycl::buffer<unsigned> hist_group{buf.size()};
    for (int bit = 0; bit < 4; bit++) {
        q.submit([&] (sycl::handler &cgh) {
            sycl::local_accessor<unsigned> count{256, cgh};
            sycl::accessor hist{hist_group, cgh, sycl::write_only, sycl::no_init};
            sycl::accessor a{buf, cgh, sycl::read_only};
            cgh.parallel_for<KT_radix_sort_histogram>(sycl::nd_range<1>{buf.size(), 256}, [=] (sycl::nd_item<1> it) {
                int ii = it.get_local_id(0);
                int gi = it.get_group(0);
                int gn = it.get_group_range(0);
                int i = it.get_global_id(0);
                count[ii] = 0;
                it.barrier(sycl::access::fence_space::local_space);
                unsigned key = (a[i] >> bit * 8) & 0xff;
                atomic_ref(count[key]).fetch_add(1u);
                it.barrier(sycl::access::fence_space::local_space);
                hist[ii * gn + gi] = count[ii];
            });
        });
        exclusive_scan(q, hist_group);
        q.submit([&] (sycl::handler &cgh) {
            sycl::local_accessor<unsigned> count{256, cgh};
            sycl::local_accessor<unsigned> bits{256 * 9, cgh};
            sycl::accessor hist{hist_group, cgh, sycl::read_only};
            sycl::accessor a{buf, cgh, sycl::read_only};
            sycl::accessor aout{buf_next, cgh, sycl::write_only, sycl::no_init};
            cgh.parallel_for<KT_radix_sort_scatter>(sycl::nd_range<1>{buf.size(), 256}, [=] (sycl::nd_item<1> it) {
                int ii = it.get_local_id(0);
                int gi = it.get_group(0);
                int gn = it.get_group_range(0);
                int i = it.get_global_id(0);
                auto g = it.get_group();
                count[ii] = hist[ii * gn + gi];
                for (int k = 0; k < 8; k++) {
                    bits[ii * 9 + k] = 0u;
                }
                it.barrier(sycl::access::fence_space::local_space);
                unsigned key = (a[i] >> bit * 8) & 0xff;
                atomic_ref(bits[key * 9 + (ii >> 5)]).fetch_or(1u << (ii & 31));
                it.barrier(sycl::access::fence_space::local_space);
                int popcorns = 0;
                for (int k = 0; k <= (ii >> 5); k++) {
                    unsigned mask = bits[key * 9 + k];
                    if (ii < (k + 1) * 32)
                        mask &= (1u << (ii & 31)) - 1u;
#if defined(__GNUC__) && defined(__OPENSYCL__)
                    popcorns += __builtin_popcount(mask); // this doesn't matter, it still fails even after commenting this out
#else
                    popcorns += sycl::popcount(mask); // not yet having this in latest develop branch
#endif
                }
                unsigned index = count[key] + popcorns;
                aout[index] = a[i];
            });
        });
        std::swap(buf, buf_next);
    }
}

int main() {
    constexpr size_t n = 65536;
    {
        sycl::queue q{sycl::gpu_selector_v};
        std::cerr << "Running on " << q.get_device().get_info<sycl::info::device::name>() << std::endl;
        std::vector<unsigned> arr(n);
        for (int i = 0; i < arr.size(); i++) {
            arr[i] = ((i * 114514) ^ 1919810);
        }
        {
            sycl::buffer<unsigned> buf{arr};
            radix_sort(q, buf);
            q.wait();
        }
        if (auto it = std::is_sorted_until(arr.begin(), arr.end()); it == arr.end()) {
            printf("sorted successfully\n");
        } else {
            printf("not sorted since %ld\n", it - arr.begin());
        }
    }
    return 0;
}
```
It generates many CU:218 errors again:
```
(opensycl) ➜  sycl2020test git:(master) ✗ syclcc hello.cpp -O3 && (sh -c 'rm -rf /tmp/*.hcf'; HIPSYCL_HCF_DUMP_DIRECTORY=/tmp ./a.out; opens
ycl-hcf-tool /tmp/*.hcf -x root images llvm-ir.global > out.bc && llvm-dis out.bc)
[hipSYCL Warning] OpenCL device NVIDIA GeForce RTX 2080 with Max-Q Design does not have a valid USM provider. Memory allocations are not pos
sible on that device.
[hipSYCL Warning] OpenCL device Intel(R) FPGA Emulation Device does not support Intel USM extensions; falling back to fine-grained system SV
M. USM pointer info queries have limited support.
Running on NVIDIA GeForce RTX 2080 with Max-Q Design
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_obje
ct: could not load module (error code = CU:218)
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_ca
che: code object creation has failed
[hipSYCL Error] from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code
 object construction failed
not sorted since 1
============== hipSYCL error report ============== 
hipSYCL has caught the following undhandled asynchronous errors: 

   0. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could 
not load module (error code = CU:218)
   1. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code 
object creation has failed
   2. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object co
nstruction failed
   3. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could 
not load module (error code = CU:218)
   4. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code 
object creation has failed
   5. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object co
nstruction failed
   6. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could 
not load module (error code = CU:218)
   7. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code 
object creation has failed
   8. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object co
nstruction failed
   9. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could 
not load module (error code = CU:218)
   10. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   11. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   12. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   13. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   14. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   15. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   16. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   17. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   18. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   19. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   20. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   21. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   22. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   23. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   24. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   25. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   26. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   27. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   28. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   29. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   30. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   31. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   32. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   33. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   34. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   35. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   36. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   37. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   38. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   39. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   40. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   41. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   42. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   43. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   44. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   45. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   46. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   47. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   48. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   49. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   50. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   51. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   52. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   53. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   54. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   55. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   56. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
   57. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_code_object.cpp:96 @ build_cuda_module_from_ptx(): cuda_executable_object: could
 not load module (error code = CU:218)
   58. from /home/bate/Codes/OpenSYCL/include/hipSYCL/runtime/kernel_cache.hpp:421 @ get_or_construct_code_object_impl(): kernel_cache: code
 object creation has failed
   59. from /home/bate/Codes/OpenSYCL/src/runtime/cuda/cuda_queue.cpp:704 @ submit_sscp_kernel_from_code_object(): cuda_queue: Code object c
onstruction failed
The application will now be terminated.
terminate called without an active exception
(opensycl) ➜  sycl2020test git:(master) ✗ 
```
IR:
```
; ModuleID = 'out.bc'
source_filename = "hello.cpp"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-linux-gnu"

%class.anon.491 = type { %"class.hipsycl::sycl::accessor", %"class.hipsycl::sycl::accessor.489", i32, %"class.hipsycl::sycl::accessor.483" }
%"class.hipsycl::sycl::accessor" = type { i64, %"class.hipsycl::sycl::range" }
%"class.hipsycl::sycl::range" = type { %"struct.hipsycl::sycl::detail::device_array" }
%"struct.hipsycl::sycl::detail::device_array" = type { i64 }
%"class.hipsycl::sycl::accessor.489" = type { %"class.hipsycl::sycl::detail::accessor_base", %"class.hipsycl::sycl::detail::conditional_storage.486" }
%"class.hipsycl::sycl::detail::accessor_base" = type { %"class.hipsycl::glue::embedded_pointer" }
%"class.hipsycl::glue::embedded_pointer" = type { [2 x ptr] }
%"class.hipsycl::sycl::detail::conditional_storage.486" = type { %"class.hipsycl::sycl::range" }
%"class.hipsycl::sycl::accessor.483" = type { %"class.hipsycl::sycl::detail::accessor_base", %"class.hipsycl::sycl::detail::conditional_storage.486" }
%"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for_offset" = type { %class.anon.491, %"struct.hipsycl::sycl::id" }
%"struct.hipsycl::sycl::id" = type { %"struct.hipsycl::sycl::detail::device_array" }
%"struct.hipsycl::sycl::group" = type { %"struct.hipsycl::sycl::id", %"class.hipsycl::sycl::range", %"class.hipsycl::sycl::range", ptr, %"struct.hipsycl::sycl::id", ptr }
%class.anon.687 = type { %"class.hipsycl::sycl::accessor.685", %"class.hipsycl::sycl::accessor.483" }
%"class.hipsycl::sycl::accessor.685" = type { %"class.hipsycl::sycl::detail::accessor_base", %"class.hipsycl::sycl::detail::conditional_storage.486" }
%"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for_offset.698" = type { %class.anon.687, %"struct.hipsycl::sycl::id" }
%"struct.hipsycl::sycl::nd_item" = type { ptr, %"struct.hipsycl::sycl::id", %"struct.hipsycl::sycl::id", %"class.hipsycl::sycl::range", %"class.hipsycl::sycl::range", %"struct.hipsycl::sycl::id", ptr, ptr }
%class.anon.765 = type { %"class.hipsycl::sycl::accessor", %"class.hipsycl::sycl::accessor.489", %"class.hipsycl::sycl::accessor", %"class.hipsycl::sycl::accessor.489", i32, %"class.hipsycl::sycl::accessor.483" }
%"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for" = type { %class.anon.491 }
%"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for.697" = type { %class.anon.687 }
%"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for.775" = type { %class.anon.765 }
%"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for_offset.776" = type { %class.anon.765, %"struct.hipsycl::sycl::id" }

$_ZNK7hipsycl4glue15__sscp_dispatch20ndrange_parallel_forIZZ10radix_sortRNS_4sycl5queueERNS3_6bufferIjLi1ESaIjEEEENKUlRNS3_7handlerEE_clESB_EUlNS3_7nd_itemILi1EEEE_Li1EEclEv = comdat any

$_ZNK7hipsycl4glue15__sscp_dispatch27ndrange_parallel_for_offsetIZZ10radix_sortRNS_4sycl5queueERNS3_6bufferIjLi1ESaIjEEEENKUlRNS3_7handlerEE_clESB_EUlNS3_7nd_itemILi1EEEE_Li1EEclEv = comdat any

$_ZNK7hipsycl4glue15__sscp_dispatch20ndrange_parallel_forIZZ14exclusive_scanRNS_4sycl5queueERNS3_6bufferIjLi1ESaIjEEEENKUlRNS3_7handlerEE_clESB_EUlNS3_7nd_itemILi1EEEE_Li1EEclEv = comdat any

$_ZNK7hipsycl4glue15__sscp_dispatch27ndrange_parallel_for_offsetIZZ14exclusive_scanRNS_4sycl5queueERNS3_6bufferIjLi1ESaIjEEEENKUlRNS3_7handlerEE_clESB_EUlNS3_7nd_itemILi1EEEE_Li1EEclEv = comdat any

$_ZZZ10radix_sortRN7hipsycl4sycl5queueERNS0_6bufferIjLi1ESaIjEEEENKUlRNS0_7handlerEE0_clES8_ENKUlNS0_7nd_itemILi1EEEE_clESB_ = comdat any

; Function Attrs: argmemonly nocallback nofree nosync nounwind willreturn
declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #0

; Function Attrs: argmemonly nocallback nofree nosync nounwind willreturn
declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #0

; Function Attrs: nounwind
define linkonce_odr dso_local void @_ZNK7hipsycl4glue15__sscp_dispatch20ndrange_parallel_forIZZ10radix_sortRNS_4sycl5queueERNS3_6bufferIjLi1ESaIjEEEENKUlRNS3_7handlerEE_clESB_EUlNS3_7nd_itemILi1EEEE_Li1EEclEv(ptr noundef nonnull align 8 dereferenceable(72) %0) local_unnamed_addr #1 comdat align 2 {
  %2 = tail call i64 @__hipsycl_sscp_get_group_id_x() #1
  %3 = tail call i64 @__hipsycl_sscp_get_local_id_x() #1
  %4 = shl i64 %3, 32
  %5 = ashr exact i64 %4, 32
  %6 = tail call i64 @__hipsycl_sscp_get_local_size_x() #1
  %7 = tail call i64 @__hipsycl_sscp_get_num_groups_x() #1
  %8 = mul i64 %6, %2
  %9 = add i64 %8, %3
  %10 = load i64, ptr %0, align 8, !tbaa !5
  %11 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %12 = addrspacecast ptr addrspace(3) %11 to ptr
  %13 = getelementptr inbounds i8, ptr %12, i64 %10
  %14 = getelementptr inbounds i32, ptr %13, i64 %5
  store i32 0, ptr %14, align 4, !tbaa !12
  tail call void @__hipsycl_sscp_work_group_barrier(i32 noundef 2, i32 noundef 4) #2
  %15 = getelementptr inbounds %class.anon.491, ptr %0, i64 0, i32 1
  %16 = shl i64 %9, 32
  %17 = ashr exact i64 %16, 32
  %18 = load ptr, ptr %15, align 8, !tbaa !14
  %19 = getelementptr inbounds i32, ptr %18, i64 %17
  %20 = load i32, ptr %19, align 4, !tbaa !12
  %21 = getelementptr inbounds %class.anon.491, ptr %0, i64 0, i32 2
  %22 = load i32, ptr %21, align 8, !tbaa !16
  %23 = shl nsw i32 %22, 3
  %24 = lshr i32 %20, %23
  %25 = and i32 %24, 255
  %26 = zext i32 %25 to i64
  %27 = load i64, ptr %0, align 8, !tbaa !5
  %28 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %29 = addrspacecast ptr addrspace(3) %28 to ptr
  %30 = getelementptr inbounds i8, ptr %29, i64 %27
  %31 = getelementptr inbounds i32, ptr %30, i64 %26
  %32 = tail call i32 @__hipsycl_sscp_atomic_fetch_add_u32(i32 noundef 4, i32 noundef 0, i32 noundef 2, ptr noundef %31, i32 noundef 1) #6
  tail call void @__hipsycl_sscp_work_group_barrier(i32 noundef 2, i32 noundef 4) #2
  %33 = load i64, ptr %0, align 8, !tbaa !5
  %34 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %35 = addrspacecast ptr addrspace(3) %34 to ptr
  %36 = getelementptr inbounds i8, ptr %35, i64 %33
  %37 = getelementptr inbounds i32, ptr %36, i64 %5
  %38 = load i32, ptr %37, align 4, !tbaa !12
  %39 = getelementptr inbounds %class.anon.491, ptr %0, i64 0, i32 3
  %40 = mul i64 %7, %3
  %41 = add i64 %40, %2
  %42 = shl i64 %41, 32
  %43 = ashr exact i64 %42, 32
  %44 = load ptr, ptr %39, align 8, !tbaa !14
  %45 = getelementptr inbounds i32, ptr %44, i64 %43
  store i32 %38, ptr %45, align 4, !tbaa !12
  ret void
}

; Function Attrs: nounwind
declare i64 @__hipsycl_sscp_get_group_id_x() local_unnamed_addr #1

; Function Attrs: nounwind
declare i64 @__hipsycl_sscp_get_local_id_x() local_unnamed_addr #1

; Function Attrs: nounwind
declare i64 @__hipsycl_sscp_get_local_size_x() local_unnamed_addr #1

; Function Attrs: nounwind
declare i64 @__hipsycl_sscp_get_num_groups_x() local_unnamed_addr #1

; Function Attrs: nounwind
declare ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() local_unnamed_addr #1

; Function Attrs: convergent nounwind
declare void @__hipsycl_sscp_work_group_barrier(i32 noundef, i32 noundef) local_unnamed_addr #2

; Function Attrs: nounwind
declare i32 @__hipsycl_sscp_atomic_fetch_add_u32(i32 noundef, i32 noundef, i32 noundef, ptr noundef, i32 noundef) local_unnamed_addr #1

; Function Attrs: nounwind
define linkonce_odr dso_local void @_ZNK7hipsycl4glue15__sscp_dispatch27ndrange_parallel_for_offsetIZZ10radix_sortRNS_4sycl5queueERNS3_6bufferIjLi1ESaIjEEEENKUlRNS3_7handlerEE_clESB_EUlNS3_7nd_itemILi1EEEE_Li1EEclEv(ptr noundef nonnull align 8 dereferenceable(80) %0) local_unnamed_addr #1 comdat align 2 {
  %2 = getelementptr inbounds %"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for_offset", ptr %0, i64 0, i32 1
  %3 = tail call i64 @__hipsycl_sscp_get_group_id_x() #1
  %4 = tail call i64 @__hipsycl_sscp_get_local_id_x() #1
  %5 = shl i64 %4, 32
  %6 = ashr exact i64 %5, 32
  %7 = tail call i64 @__hipsycl_sscp_get_local_size_x() #1
  %8 = tail call i64 @__hipsycl_sscp_get_num_groups_x() #1
  %9 = mul i64 %7, %3
  %10 = add i64 %9, %4
  %11 = load i64, ptr %2, align 8, !tbaa !23
  %12 = add i64 %10, %11
  %13 = load i64, ptr %0, align 8, !tbaa !5
  %14 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %15 = addrspacecast ptr addrspace(3) %14 to ptr
  %16 = getelementptr inbounds i8, ptr %15, i64 %13
  %17 = getelementptr inbounds i32, ptr %16, i64 %6
  store i32 0, ptr %17, align 4, !tbaa !12
  tail call void @__hipsycl_sscp_work_group_barrier(i32 noundef 2, i32 noundef 4) #2
  %18 = getelementptr inbounds %class.anon.491, ptr %0, i64 0, i32 1
  %19 = shl i64 %12, 32
  %20 = ashr exact i64 %19, 32
  %21 = load ptr, ptr %18, align 8, !tbaa !14
  %22 = getelementptr inbounds i32, ptr %21, i64 %20
  %23 = load i32, ptr %22, align 4, !tbaa !12
  %24 = getelementptr inbounds %class.anon.491, ptr %0, i64 0, i32 2
  %25 = load i32, ptr %24, align 8, !tbaa !16
  %26 = shl nsw i32 %25, 3
  %27 = lshr i32 %23, %26
  %28 = and i32 %27, 255
  %29 = zext i32 %28 to i64
  %30 = load i64, ptr %0, align 8, !tbaa !5
  %31 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %32 = addrspacecast ptr addrspace(3) %31 to ptr
  %33 = getelementptr inbounds i8, ptr %32, i64 %30
  %34 = getelementptr inbounds i32, ptr %33, i64 %29
  %35 = tail call i32 @__hipsycl_sscp_atomic_fetch_add_u32(i32 noundef 4, i32 noundef 0, i32 noundef 2, ptr noundef %34, i32 noundef 1) #6
  tail call void @__hipsycl_sscp_work_group_barrier(i32 noundef 2, i32 noundef 4) #2
  %36 = load i64, ptr %0, align 8, !tbaa !5
  %37 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %38 = addrspacecast ptr addrspace(3) %37 to ptr
  %39 = getelementptr inbounds i8, ptr %38, i64 %36
  %40 = getelementptr inbounds i32, ptr %39, i64 %6
  %41 = load i32, ptr %40, align 4, !tbaa !12
  %42 = getelementptr inbounds %class.anon.491, ptr %0, i64 0, i32 3
  %43 = mul i64 %8, %4
  %44 = add i64 %43, %3
  %45 = shl i64 %44, 32
  %46 = ashr exact i64 %45, 32
  %47 = load ptr, ptr %42, align 8, !tbaa !14
  %48 = getelementptr inbounds i32, ptr %47, i64 %46
  store i32 %41, ptr %48, align 4, !tbaa !12
  ret void
}

; Function Attrs: nounwind
define linkonce_odr dso_local void @_ZNK7hipsycl4glue15__sscp_dispatch20ndrange_parallel_forIZZ14exclusive_scanRNS_4sycl5queueERNS3_6bufferIjLi1ESaIjEEEENKUlRNS3_7handlerEE_clESB_EUlNS3_7nd_itemILi1EEEE_Li1EEclEv(ptr noundef nonnull align 8 dereferenceable(48) %0) local_unnamed_addr #1 comdat align 2 {
  %2 = alloca %"struct.hipsycl::sycl::group", align 8
  %3 = tail call i64 @__hipsycl_sscp_get_group_id_x() #1
  %4 = shl i64 %3, 32
  %5 = ashr exact i64 %4, 32
  %6 = tail call i64 @__hipsycl_sscp_get_local_id_x() #1
  %7 = shl i64 %6, 32
  %8 = ashr exact i64 %7, 32
  %9 = tail call i64 @__hipsycl_sscp_get_local_size_x() #1
  %10 = shl i64 %9, 32
  %11 = ashr exact i64 %10, 32
  %12 = tail call i64 @__hipsycl_sscp_get_num_groups_x() #1
  %13 = shl i64 %12, 32
  %14 = ashr exact i64 %13, 32
  %15 = mul i64 %9, %3
  %16 = add i64 %15, %6
  %17 = trunc i64 %6 to i32
  %18 = shl i64 %16, 32
  %19 = ashr exact i64 %18, 32
  %20 = getelementptr inbounds i8, ptr %0, i64 16
  %21 = load i64, ptr %20, align 8, !tbaa.struct !24
  %22 = icmp ult i64 %19, %21
  br i1 %22, label %23, label %27

23:                                               ; preds = %1
  %24 = load ptr, ptr %0, align 8, !tbaa !14
  %25 = getelementptr inbounds i32, ptr %24, i64 %19
  %26 = load i32, ptr %25, align 4, !tbaa !12
  br label %27

27:                                               ; preds = %23, %1
  %28 = phi i32 [ %26, %23 ], [ 0, %1 ]
  call void @llvm.lifetime.start.p0(i64 48, ptr nonnull %2)
  store i64 %5, ptr %2, align 8
  %29 = getelementptr inbounds i8, ptr %2, i64 8
  store i64 %11, ptr %29, align 8
  %30 = getelementptr inbounds i8, ptr %2, i64 16
  store i64 %14, ptr %30, align 8
  %31 = getelementptr inbounds i8, ptr %2, i64 32
  store i64 %8, ptr %31, align 8
  %32 = getelementptr inbounds i8, ptr %2, i64 40
  store ptr null, ptr %32, align 8
  %33 = tail call noundef i32 @_ZN7hipsycl4sycl6detail13sscp_builtins35__hipsycl_inclusive_scan_over_groupILi1EjSt4plusIvEEET0_NS0_5groupIXT_EEES6_T1_(ptr noundef nonnull byval(%"struct.hipsycl::sycl::group") align 8 %2, i32 noundef %28) #6
  call void @llvm.lifetime.end.p0(i64 48, ptr nonnull %2)
  %34 = icmp eq i32 %17, 255
  br i1 %34, label %35, label %37

35:                                               ; preds = %27
  %36 = getelementptr inbounds %class.anon.687, ptr %0, i64 0, i32 1
  br label %47

37:                                               ; preds = %27
  %38 = add i64 %18, 4294967296
  %39 = ashr exact i64 %38, 32
  %40 = load i64, ptr %20, align 8, !tbaa.struct !24
  %41 = icmp ult i64 %39, %40
  br i1 %41, label %42, label %45

42:                                               ; preds = %37
  %43 = load ptr, ptr %0, align 8, !tbaa !14
  %44 = getelementptr inbounds i32, ptr %43, i64 %39
  store i32 %33, ptr %44, align 4, !tbaa !12
  br label %45

45:                                               ; preds = %42, %37
  %46 = icmp eq i32 %17, 0
  br i1 %46, label %47, label %53

47:                                               ; preds = %45, %35
  %48 = phi ptr [ %36, %35 ], [ %0, %45 ]
  %49 = phi i64 [ %5, %35 ], [ %19, %45 ]
  %50 = phi i32 [ %33, %35 ], [ 0, %45 ]
  %51 = load ptr, ptr %48, align 8, !tbaa !14
  %52 = getelementptr inbounds i32, ptr %51, i64 %49
  store i32 %50, ptr %52, align 4, !tbaa !12
  br label %53

53:                                               ; preds = %47, %45
  ret void
}

; Function Attrs: nounwind
declare noundef i32 @_ZN7hipsycl4sycl6detail13sscp_builtins35__hipsycl_inclusive_scan_over_groupILi1EjSt4plusIvEEET0_NS0_5groupIXT_EEES6_T1_(ptr noundef byval(%"struct.hipsycl::sycl::group") align 8, i32 noundef) local_unnamed_addr #1

; Function Attrs: nounwind
define linkonce_odr dso_local void @_ZNK7hipsycl4glue15__sscp_dispatch27ndrange_parallel_for_offsetIZZ14exclusive_scanRNS_4sycl5queueERNS3_6bufferIjLi1ESaIjEEEENKUlRNS3_7handlerEE_clESB_EUlNS3_7nd_itemILi1EEEE_Li1EEclEv(ptr noundef nonnull align 8 dereferenceable(56) %0) local_unnamed_addr #1 comdat align 2 {
  %2 = alloca %"struct.hipsycl::sycl::group", align 8
  %3 = getelementptr inbounds %"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for_offset.698", ptr %0, i64 0, i32 1
  %4 = tail call i64 @__hipsycl_sscp_get_group_id_x() #1
  %5 = shl i64 %4, 32
  %6 = ashr exact i64 %5, 32
  %7 = tail call i64 @__hipsycl_sscp_get_local_id_x() #1
  %8 = shl i64 %7, 32
  %9 = ashr exact i64 %8, 32
  %10 = tail call i64 @__hipsycl_sscp_get_local_size_x() #1
  %11 = shl i64 %10, 32
  %12 = ashr exact i64 %11, 32
  %13 = tail call i64 @__hipsycl_sscp_get_num_groups_x() #1
  %14 = shl i64 %13, 32
  %15 = ashr exact i64 %14, 32
  %16 = mul i64 %10, %4
  %17 = add i64 %16, %7
  %18 = trunc i64 %7 to i32
  %19 = load i64, ptr %3, align 8, !tbaa !23
  %20 = add i64 %17, %19
  %21 = shl i64 %20, 32
  %22 = ashr exact i64 %21, 32
  %23 = getelementptr inbounds i8, ptr %0, i64 16
  %24 = load i64, ptr %23, align 8, !tbaa.struct !24
  %25 = icmp ult i64 %22, %24
  br i1 %25, label %26, label %30

26:                                               ; preds = %1
  %27 = load ptr, ptr %0, align 8, !tbaa !14
  %28 = getelementptr inbounds i32, ptr %27, i64 %22
  %29 = load i32, ptr %28, align 4, !tbaa !12
  br label %30

30:                                               ; preds = %26, %1
  %31 = phi i32 [ %29, %26 ], [ 0, %1 ]
  call void @llvm.lifetime.start.p0(i64 48, ptr nonnull %2)
  store i64 %6, ptr %2, align 8
  %32 = getelementptr inbounds i8, ptr %2, i64 8
  store i64 %12, ptr %32, align 8
  %33 = getelementptr inbounds i8, ptr %2, i64 16
  store i64 %15, ptr %33, align 8
  %34 = getelementptr inbounds i8, ptr %2, i64 32
  store i64 %9, ptr %34, align 8
  %35 = getelementptr inbounds i8, ptr %2, i64 40
  store ptr null, ptr %35, align 8
  %36 = tail call noundef i32 @_ZN7hipsycl4sycl6detail13sscp_builtins35__hipsycl_inclusive_scan_over_groupILi1EjSt4plusIvEEET0_NS0_5groupIXT_EEES6_T1_(ptr noundef nonnull byval(%"struct.hipsycl::sycl::group") align 8 %2, i32 noundef %31) #6
  call void @llvm.lifetime.end.p0(i64 48, ptr nonnull %2)
  %37 = icmp eq i32 %18, 255
  br i1 %37, label %38, label %40

38:                                               ; preds = %30
  %39 = getelementptr inbounds %class.anon.687, ptr %0, i64 0, i32 1
  br label %50

40:                                               ; preds = %30
  %41 = add i64 %21, 4294967296
  %42 = ashr exact i64 %41, 32
  %43 = load i64, ptr %23, align 8, !tbaa.struct !24
  %44 = icmp ult i64 %42, %43
  br i1 %44, label %45, label %48

45:                                               ; preds = %40
  %46 = load ptr, ptr %0, align 8, !tbaa !14
  %47 = getelementptr inbounds i32, ptr %46, i64 %42
  store i32 %36, ptr %47, align 4, !tbaa !12
  br label %48

48:                                               ; preds = %45, %40
  %49 = icmp eq i32 %18, 0
  br i1 %49, label %50, label %56

50:                                               ; preds = %48, %38
  %51 = phi ptr [ %39, %38 ], [ %0, %48 ]
  %52 = phi i64 [ %6, %38 ], [ %22, %48 ]
  %53 = phi i32 [ %36, %38 ], [ 0, %48 ]
  %54 = load ptr, ptr %51, align 8, !tbaa !14
  %55 = getelementptr inbounds i32, ptr %54, i64 %52
  store i32 %53, ptr %55, align 4, !tbaa !12
  br label %56

56:                                               ; preds = %50, %48
  ret void
}

; Function Attrs: inlinehint nounwind
define linkonce_odr dso_local void @_ZZZ10radix_sortRN7hipsycl4sycl5queueERNS0_6bufferIjLi1ESaIjEEEENKUlRNS0_7handlerEE0_clES8_ENKUlNS0_7nd_itemILi1EEEE_clESB_(ptr noundef nonnull align 8 dereferenceable(112) %0, ptr noundef byval(%"struct.hipsycl::sycl::nd_item") align 8 %1) local_unnamed_addr #3 comdat align 2 {
  %3 = getelementptr inbounds %"struct.hipsycl::sycl::nd_item", ptr %1, i64 0, i32 2
  %4 = load i64, ptr %3, align 8, !tbaa !23
  %5 = trunc i64 %4 to i32
  %6 = getelementptr inbounds %"struct.hipsycl::sycl::nd_item", ptr %1, i64 0, i32 1
  %7 = load i64, ptr %6, align 8, !tbaa !23
  %8 = getelementptr inbounds %"struct.hipsycl::sycl::nd_item", ptr %1, i64 0, i32 4
  %9 = load i64, ptr %8, align 8, !tbaa !23
  %10 = getelementptr inbounds %"struct.hipsycl::sycl::nd_item", ptr %1, i64 0, i32 5
  %11 = load i64, ptr %10, align 8, !tbaa !23
  %12 = load ptr, ptr %1, align 8, !tbaa !25
  %13 = load i64, ptr %12, align 8, !tbaa !23
  %14 = getelementptr inbounds %class.anon.765, ptr %0, i64 0, i32 1
  %15 = mul i64 %9, %4
  %16 = add i64 %15, %7
  %17 = shl i64 %16, 32
  %18 = ashr exact i64 %17, 32
  %19 = load ptr, ptr %14, align 8, !tbaa !14
  %20 = getelementptr inbounds i32, ptr %19, i64 %18
  %21 = load i32, ptr %20, align 4, !tbaa !12
  %22 = shl i64 %4, 32
  %23 = ashr exact i64 %22, 32
  %24 = load i64, ptr %0, align 8, !tbaa !5
  %25 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %26 = addrspacecast ptr addrspace(3) %25 to ptr
  %27 = getelementptr inbounds i8, ptr %26, i64 %24
  %28 = getelementptr inbounds i32, ptr %27, i64 %23
  store i32 %21, ptr %28, align 4, !tbaa !12
  %29 = getelementptr inbounds %class.anon.765, ptr %0, i64 0, i32 2
  %30 = mul i64 %4, 38654705664
  %31 = ashr exact i64 %30, 32
  %32 = load i64, ptr %29, align 8, !tbaa !5
  %33 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %34 = addrspacecast ptr addrspace(3) %33 to ptr
  %35 = getelementptr inbounds i8, ptr %34, i64 %32
  %36 = getelementptr inbounds i32, ptr %35, i64 %31
  store i32 0, ptr %36, align 4, !tbaa !12
  %37 = add nsw i64 %31, 1
  %38 = load i64, ptr %29, align 8, !tbaa !5
  %39 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %40 = addrspacecast ptr addrspace(3) %39 to ptr
  %41 = getelementptr inbounds i8, ptr %40, i64 %38
  %42 = getelementptr inbounds i32, ptr %41, i64 %37
  store i32 0, ptr %42, align 4, !tbaa !12
  %43 = add nsw i64 %31, 2
  %44 = load i64, ptr %29, align 8, !tbaa !5
  %45 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %46 = addrspacecast ptr addrspace(3) %45 to ptr
  %47 = getelementptr inbounds i8, ptr %46, i64 %44
  %48 = getelementptr inbounds i32, ptr %47, i64 %43
  store i32 0, ptr %48, align 4, !tbaa !12
  %49 = add nsw i64 %31, 3
  %50 = load i64, ptr %29, align 8, !tbaa !5
  %51 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %52 = addrspacecast ptr addrspace(3) %51 to ptr
  %53 = getelementptr inbounds i8, ptr %52, i64 %50
  %54 = getelementptr inbounds i32, ptr %53, i64 %49
  store i32 0, ptr %54, align 4, !tbaa !12
  %55 = add nsw i64 %31, 4
  %56 = load i64, ptr %29, align 8, !tbaa !5
  %57 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %58 = addrspacecast ptr addrspace(3) %57 to ptr
  %59 = getelementptr inbounds i8, ptr %58, i64 %56
  %60 = getelementptr inbounds i32, ptr %59, i64 %55
  store i32 0, ptr %60, align 4, !tbaa !12
  %61 = add nsw i64 %31, 5
  %62 = load i64, ptr %29, align 8, !tbaa !5
  %63 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %64 = addrspacecast ptr addrspace(3) %63 to ptr
  %65 = getelementptr inbounds i8, ptr %64, i64 %62
  %66 = getelementptr inbounds i32, ptr %65, i64 %61
  store i32 0, ptr %66, align 4, !tbaa !12
  %67 = add nsw i64 %31, 6
  %68 = load i64, ptr %29, align 8, !tbaa !5
  %69 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %70 = addrspacecast ptr addrspace(3) %69 to ptr
  %71 = getelementptr inbounds i8, ptr %70, i64 %68
  %72 = getelementptr inbounds i32, ptr %71, i64 %67
  store i32 0, ptr %72, align 4, !tbaa !12
  %73 = add nsw i64 %31, 7
  %74 = load i64, ptr %29, align 8, !tbaa !5
  %75 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %76 = addrspacecast ptr addrspace(3) %75 to ptr
  %77 = getelementptr inbounds i8, ptr %76, i64 %74
  %78 = getelementptr inbounds i32, ptr %77, i64 %73
  store i32 0, ptr %78, align 4, !tbaa !12
  %79 = add i64 %13, %11
  tail call void @__hipsycl_sscp_work_group_barrier(i32 noundef 2, i32 noundef 4) #2
  %80 = getelementptr inbounds %class.anon.765, ptr %0, i64 0, i32 3
  %81 = shl i64 %79, 32
  %82 = ashr exact i64 %81, 32
  %83 = load ptr, ptr %80, align 8, !tbaa !14
  %84 = getelementptr inbounds i32, ptr %83, i64 %82
  %85 = load i32, ptr %84, align 4, !tbaa !12
  %86 = getelementptr inbounds %class.anon.765, ptr %0, i64 0, i32 4
  %87 = load i32, ptr %86, align 8, !tbaa !28
  %88 = shl nsw i32 %87, 3
  %89 = lshr i32 %85, %88
  %90 = and i32 %89, 255
  %91 = mul nuw nsw i32 %90, 9
  %92 = ashr i32 %5, 5
  %93 = add nsw i32 %91, %92
  %94 = zext i32 %93 to i64
  %95 = load i64, ptr %29, align 8, !tbaa !5
  %96 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %97 = addrspacecast ptr addrspace(3) %96 to ptr
  %98 = getelementptr inbounds i8, ptr %97, i64 %95
  %99 = getelementptr inbounds i32, ptr %98, i64 %94
  %100 = and i32 %5, 31
  %101 = shl nuw i32 1, %100
  %102 = tail call i32 @__hipsycl_sscp_atomic_fetch_or_i32(i32 noundef 4, i32 noundef 0, i32 noundef 2, ptr noundef %99, i32 noundef %101) #6
  tail call void @__hipsycl_sscp_work_group_barrier(i32 noundef 2, i32 noundef 4) #2
  %103 = icmp slt i32 %5, 0
  br i1 %103, label %129, label %104

104:                                              ; preds = %2
  %105 = add i32 %101, -1
  %106 = zext i32 %91 to i64
  %107 = tail call i32 @llvm.smax.i32(i32 %92, i32 0)
  %108 = add nuw nsw i32 %107, 1
  %109 = zext i32 %108 to i64
  br label %110

110:                                              ; preds = %110, %104
  %111 = phi i64 [ 0, %104 ], [ %120, %110 ]
  %112 = phi i32 [ 0, %104 ], [ %127, %110 ]
  %113 = add nuw nsw i64 %111, %106
  %114 = load i64, ptr %29, align 8, !tbaa !5
  %115 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %116 = addrspacecast ptr addrspace(3) %115 to ptr
  %117 = getelementptr inbounds i8, ptr %116, i64 %114
  %118 = getelementptr inbounds i32, ptr %117, i64 %113
  %119 = load i32, ptr %118, align 4, !tbaa !12
  %120 = add nuw nsw i64 %111, 1
  %121 = trunc i64 %120 to i32
  %122 = shl i32 %121, 5
  %123 = icmp sgt i32 %122, %5
  %124 = select i1 %123, i32 %105, i32 -1
  %125 = and i32 %119, %124
  %126 = tail call i32 @llvm.ctpop.i32(i32 %125), !range !30
  %127 = add nuw nsw i32 %126, %112
  %128 = icmp eq i64 %120, %109
  br i1 %128, label %129, label %110, !llvm.loop !31

129:                                              ; preds = %110, %2
  %130 = phi i32 [ 0, %2 ], [ %127, %110 ]
  %131 = zext i32 %90 to i64
  %132 = load i64, ptr %0, align 8, !tbaa !5
  %133 = tail call ptr addrspace(3) @__hipsycl_sscp_get_dynamic_local_memory() #1
  %134 = addrspacecast ptr addrspace(3) %133 to ptr
  %135 = getelementptr inbounds i8, ptr %134, i64 %132
  %136 = getelementptr inbounds i32, ptr %135, i64 %131
  %137 = load i32, ptr %136, align 4, !tbaa !12
  %138 = add i32 %137, %130
  %139 = load ptr, ptr %80, align 8, !tbaa !14
  %140 = getelementptr inbounds i32, ptr %139, i64 %82
  %141 = load i32, ptr %140, align 4, !tbaa !12
  %142 = getelementptr inbounds %class.anon.765, ptr %0, i64 0, i32 5
  %143 = zext i32 %138 to i64
  %144 = load ptr, ptr %142, align 8, !tbaa !14
  %145 = getelementptr inbounds i32, ptr %144, i64 %143
  store i32 %141, ptr %145, align 4, !tbaa !12
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind readnone speculatable willreturn
declare i32 @llvm.ctpop.i32(i32) #4

; Function Attrs: nounwind
declare i32 @__hipsycl_sscp_atomic_fetch_or_i32(i32 noundef, i32 noundef, i32 noundef, ptr noundef, i32 noundef) local_unnamed_addr #1

; Function Attrs: mustprogress nounwind
define void @_Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch20ndrange_parallel_forIZZ10radix_sortRNS0_4sycl5queueERNS4_6bufferIjLi1ESaIjEEEENKUlRNS4_7handlerEE_clESC_EUlNS4_7nd_itemILi1EEEE_Li1EEEEvRKT_(i64 %0, i64 %1, ptr %2, ptr %3, i64 %4, i32 %5, ptr %6, ptr %7, i64 %8) local_unnamed_addr #5 {
  %10 = alloca %"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for", align 8
  call void @llvm.lifetime.start.p0(i64 72, ptr nonnull %10) #1
  store i64 %0, ptr %10, align 8
  %11 = getelementptr inbounds i8, ptr %10, i64 8
  store i64 %1, ptr %11, align 8
  %12 = getelementptr inbounds i8, ptr %10, i64 16
  store ptr %2, ptr %12, align 8
  %13 = getelementptr inbounds i8, ptr %10, i64 24
  store ptr %3, ptr %13, align 8
  %14 = getelementptr inbounds i8, ptr %10, i64 32
  store i64 %4, ptr %14, align 8
  %15 = getelementptr inbounds i8, ptr %10, i64 40
  store i32 %5, ptr %15, align 8
  %16 = getelementptr inbounds i8, ptr %10, i64 48
  store ptr %6, ptr %16, align 8
  %17 = getelementptr inbounds i8, ptr %10, i64 56
  store ptr %7, ptr %17, align 8
  %18 = getelementptr inbounds i8, ptr %10, i64 64
  store i64 %8, ptr %18, align 8
  call void @_ZNK7hipsycl4glue15__sscp_dispatch20ndrange_parallel_forIZZ10radix_sortRNS_4sycl5queueERNS3_6bufferIjLi1ESaIjEEEENKUlRNS3_7handlerEE_clESB_EUlNS3_7nd_itemILi1EEEE_Li1EEclEv(ptr noundef nonnull align 8 dereferenceable(72) %10) #1
  call void @llvm.lifetime.end.p0(i64 72, ptr nonnull %10) #1
  ret void
}

; Function Attrs: mustprogress nounwind
define void @_Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch27ndrange_parallel_for_offsetIZZ10radix_sortRNS0_4sycl5queueERNS4_6bufferIjLi1ESaIjEEEENKUlRNS4_7handlerEE_clESC_EUlNS4_7nd_itemILi1EEEE_Li1EEEEvRKT_(i64 %0, i64 %1, ptr %2, ptr %3, i64 %4, i32 %5, ptr %6, ptr %7, i64 %8, i64 %9) local_unnamed_addr #5 {
  %11 = alloca %"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for_offset", align 8
  call void @llvm.lifetime.start.p0(i64 80, ptr nonnull %11) #1
  store i64 %0, ptr %11, align 8
  %12 = getelementptr inbounds i8, ptr %11, i64 8
  store i64 %1, ptr %12, align 8
  %13 = getelementptr inbounds i8, ptr %11, i64 16
  store ptr %2, ptr %13, align 8
  %14 = getelementptr inbounds i8, ptr %11, i64 24
  store ptr %3, ptr %14, align 8
  %15 = getelementptr inbounds i8, ptr %11, i64 32
  store i64 %4, ptr %15, align 8
  %16 = getelementptr inbounds i8, ptr %11, i64 40
  store i32 %5, ptr %16, align 8
  %17 = getelementptr inbounds i8, ptr %11, i64 48
  store ptr %6, ptr %17, align 8
  %18 = getelementptr inbounds i8, ptr %11, i64 56
  store ptr %7, ptr %18, align 8
  %19 = getelementptr inbounds i8, ptr %11, i64 64
  store i64 %8, ptr %19, align 8
  %20 = getelementptr inbounds i8, ptr %11, i64 72
  store i64 %9, ptr %20, align 8
  call void @_ZNK7hipsycl4glue15__sscp_dispatch27ndrange_parallel_for_offsetIZZ10radix_sortRNS_4sycl5queueERNS3_6bufferIjLi1ESaIjEEEENKUlRNS3_7handlerEE_clESB_EUlNS3_7nd_itemILi1EEEE_Li1EEclEv(ptr noundef nonnull align 8 dereferenceable(80) %11) #1
  call void @llvm.lifetime.end.p0(i64 80, ptr nonnull %11) #1
  ret void
}

; Function Attrs: mustprogress nounwind
define void @_Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch20ndrange_parallel_forIZZ14exclusive_scanRNS0_4sycl5queueERNS4_6bufferIjLi1ESaIjEEEENKUlRNS4_7handlerEE_clESC_EUlNS4_7nd_itemILi1EEEE_Li1EEEEvRKT_(ptr %0, ptr %1, i64 %2, ptr %3, ptr %4, i64 %5) local_unnamed_addr #5 {
  %7 = alloca %"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for.697", align 8
  call void @llvm.lifetime.start.p0(i64 48, ptr nonnull %7) #1
  store ptr %0, ptr %7, align 8
  %8 = getelementptr inbounds i8, ptr %7, i64 8
  store ptr %1, ptr %8, align 8
  %9 = getelementptr inbounds i8, ptr %7, i64 16
  store i64 %2, ptr %9, align 8
  %10 = getelementptr inbounds i8, ptr %7, i64 24
  store ptr %3, ptr %10, align 8
  %11 = getelementptr inbounds i8, ptr %7, i64 32
  store ptr %4, ptr %11, align 8
  %12 = getelementptr inbounds i8, ptr %7, i64 40
  store i64 %5, ptr %12, align 8
  call void @_ZNK7hipsycl4glue15__sscp_dispatch20ndrange_parallel_forIZZ14exclusive_scanRNS_4sycl5queueERNS3_6bufferIjLi1ESaIjEEEENKUlRNS3_7handlerEE_clESB_EUlNS3_7nd_itemILi1EEEE_Li1EEclEv(ptr noundef nonnull align 8 dereferenceable(48) %7) #1
  call void @llvm.lifetime.end.p0(i64 48, ptr nonnull %7) #1
  ret void
}

; Function Attrs: mustprogress nounwind
define void @_Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch27ndrange_parallel_for_offsetIZZ14exclusive_scanRNS0_4sycl5queueERNS4_6bufferIjLi1ESaIjEEEENKUlRNS4_7handlerEE_clESC_EUlNS4_7nd_itemILi1EEEE_Li1EEEEvRKT_(ptr %0, ptr %1, i64 %2, ptr %3, ptr %4, i64 %5, i64 %6) local_unnamed_addr #5 {
  %8 = alloca %"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for_offset.698", align 8
  call void @llvm.lifetime.start.p0(i64 56, ptr nonnull %8) #1
  store ptr %0, ptr %8, align 8
  %9 = getelementptr inbounds i8, ptr %8, i64 8
  store ptr %1, ptr %9, align 8
  %10 = getelementptr inbounds i8, ptr %8, i64 16
  store i64 %2, ptr %10, align 8
  %11 = getelementptr inbounds i8, ptr %8, i64 24
  store ptr %3, ptr %11, align 8
  %12 = getelementptr inbounds i8, ptr %8, i64 32
  store ptr %4, ptr %12, align 8
  %13 = getelementptr inbounds i8, ptr %8, i64 40
  store i64 %5, ptr %13, align 8
  %14 = getelementptr inbounds i8, ptr %8, i64 48
  store i64 %6, ptr %14, align 8
  call void @_ZNK7hipsycl4glue15__sscp_dispatch27ndrange_parallel_for_offsetIZZ14exclusive_scanRNS_4sycl5queueERNS3_6bufferIjLi1ESaIjEEEENKUlRNS3_7handlerEE_clESB_EUlNS3_7nd_itemILi1EEEE_Li1EEclEv(ptr noundef nonnull align 8 dereferenceable(56) %8) #1
  call void @llvm.lifetime.end.p0(i64 56, ptr nonnull %8) #1
  ret void
}

; Function Attrs: nounwind
define void @_Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch20ndrange_parallel_forIZZ14exclusive_scanRNS0_4sycl5queueERNS4_6bufferIjLi1ESaIjEEEENKUlRNS4_7handlerEE0_clESC_EUlNS4_7nd_itemILi1EEEE_Li1EEEEvRKT_(ptr %0, ptr %1, i64 %2, ptr %3, ptr %4, i64 %5) local_unnamed_addr #1 {
  %7 = tail call i64 @__hipsycl_sscp_get_group_id_x() #1
  %8 = tail call i64 @__hipsycl_sscp_get_local_id_x() #1
  %9 = tail call i64 @__hipsycl_sscp_get_local_size_x() #1
  %10 = tail call i64 @__hipsycl_sscp_get_num_groups_x() #1
  %11 = mul i64 %9, %7
  %12 = add i64 %11, %8
  %13 = shl i64 %12, 32
  %14 = ashr exact i64 %13, 32
  %15 = icmp ult i64 %14, %2
  br i1 %15, label %16, label %24

16:                                               ; preds = %6
  %17 = shl i64 %7, 32
  %18 = ashr exact i64 %17, 32
  %19 = getelementptr inbounds i32, ptr %3, i64 %18
  %20 = load i32, ptr %19, align 4, !tbaa !12
  %21 = getelementptr inbounds i32, ptr %0, i64 %14
  %22 = load i32, ptr %21, align 4, !tbaa !12
  %23 = add i32 %22, %20
  store i32 %23, ptr %21, align 4, !tbaa !12
  br label %24

24:                                               ; preds = %16, %6
  ret void
}

; Function Attrs: nounwind
define void @_Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch27ndrange_parallel_for_offsetIZZ14exclusive_scanRNS0_4sycl5queueERNS4_6bufferIjLi1ESaIjEEEENKUlRNS4_7handlerEE0_clESC_EUlNS4_7nd_itemILi1EEEE_Li1EEEEvRKT_(ptr %0, ptr %1, i64 %2, ptr %3, ptr %4, i64 %5, i64 %6) local_unnamed_addr #1 {
  %8 = tail call i64 @__hipsycl_sscp_get_group_id_x() #1
  %9 = tail call i64 @__hipsycl_sscp_get_local_id_x() #1
  %10 = tail call i64 @__hipsycl_sscp_get_local_size_x() #1
  %11 = tail call i64 @__hipsycl_sscp_get_num_groups_x() #1
  %12 = mul i64 %10, %8
  %13 = add i64 %9, %6
  %14 = add i64 %13, %12
  %15 = shl i64 %14, 32
  %16 = ashr exact i64 %15, 32
  %17 = icmp ult i64 %16, %2
  br i1 %17, label %18, label %26

18:                                               ; preds = %7
  %19 = shl i64 %8, 32
  %20 = ashr exact i64 %19, 32
  %21 = getelementptr inbounds i32, ptr %3, i64 %20
  %22 = load i32, ptr %21, align 4, !tbaa !12
  %23 = getelementptr inbounds i32, ptr %0, i64 %16
  %24 = load i32, ptr %23, align 4, !tbaa !12
  %25 = add i32 %24, %22
  store i32 %25, ptr %23, align 4, !tbaa !12
  br label %26

26:                                               ; preds = %18, %7
  ret void
}

; Function Attrs: nounwind
define void @_Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch20ndrange_parallel_forIZZ10radix_sortRNS0_4sycl5queueERNS4_6bufferIjLi1ESaIjEEEENKUlRNS4_7handlerEE0_clESC_EUlNS4_7nd_itemILi1EEEE_Li1EEEEvRKT_(i64 %0, i64 %1, ptr %2, ptr %3, i64 %4, i64 %5, i64 %6, ptr %7, ptr %8, i64 %9, i32 %10, ptr %11, ptr %12, i64 %13) local_unnamed_addr #1 {
  %15 = alloca %"struct.hipsycl::sycl::id", align 8
  %16 = alloca %"struct.hipsycl::sycl::nd_item", align 8
  %17 = alloca %"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for.775", align 8
  call void @llvm.lifetime.start.p0(i64 112, ptr nonnull %17) #1
  store i64 %0, ptr %17, align 8
  %18 = getelementptr inbounds i8, ptr %17, i64 8
  store i64 %1, ptr %18, align 8
  %19 = getelementptr inbounds i8, ptr %17, i64 16
  store ptr %2, ptr %19, align 8
  %20 = getelementptr inbounds i8, ptr %17, i64 24
  store ptr %3, ptr %20, align 8
  %21 = getelementptr inbounds i8, ptr %17, i64 32
  store i64 %4, ptr %21, align 8
  %22 = getelementptr inbounds i8, ptr %17, i64 40
  store i64 %5, ptr %22, align 8
  %23 = getelementptr inbounds i8, ptr %17, i64 48
  store i64 %6, ptr %23, align 8
  %24 = getelementptr inbounds i8, ptr %17, i64 56
  store ptr %7, ptr %24, align 8
  %25 = getelementptr inbounds i8, ptr %17, i64 64
  store ptr %8, ptr %25, align 8
  %26 = getelementptr inbounds i8, ptr %17, i64 72
  store i64 %9, ptr %26, align 8
  %27 = getelementptr inbounds i8, ptr %17, i64 80
  store i32 %10, ptr %27, align 8
  %28 = getelementptr inbounds i8, ptr %17, i64 88
  store ptr %11, ptr %28, align 8
  %29 = getelementptr inbounds i8, ptr %17, i64 96
  store ptr %12, ptr %29, align 8
  %30 = getelementptr inbounds i8, ptr %17, i64 104
  store i64 %13, ptr %30, align 8
  call void @llvm.lifetime.start.p0(i64 64, ptr nonnull %16)
  call void @llvm.lifetime.start.p0(i64 8, ptr nonnull %15) #1
  store i64 0, ptr %15, align 8, !tbaa !33
  %31 = tail call i64 @__hipsycl_sscp_get_group_id_x() #1
  %32 = shl i64 %31, 32
  %33 = ashr exact i64 %32, 32
  %34 = tail call i64 @__hipsycl_sscp_get_local_id_x() #1
  %35 = shl i64 %34, 32
  %36 = ashr exact i64 %35, 32
  %37 = tail call i64 @__hipsycl_sscp_get_local_size_x() #1
  %38 = shl i64 %37, 32
  %39 = ashr exact i64 %38, 32
  %40 = tail call i64 @__hipsycl_sscp_get_num_groups_x() #1
  %41 = shl i64 %40, 32
  %42 = ashr exact i64 %41, 32
  %43 = mul nsw i64 %39, %33
  %44 = add nsw i64 %43, %36
  store ptr %15, ptr %16, align 8, !tbaa.struct !34
  %45 = getelementptr inbounds i8, ptr %16, i64 8
  store i64 %33, ptr %45, align 8, !tbaa.struct !35
  %46 = getelementptr inbounds i8, ptr %16, i64 16
  store i64 %36, ptr %46, align 8, !tbaa.struct !36
  %47 = getelementptr inbounds i8, ptr %16, i64 24
  store i64 %39, ptr %47, align 8, !tbaa.struct !37
  %48 = getelementptr inbounds i8, ptr %16, i64 32
  store i64 %42, ptr %48, align 8, !tbaa.struct !38
  %49 = getelementptr inbounds i8, ptr %16, i64 40
  store i64 %44, ptr %49, align 8, !tbaa.struct !39
  %50 = getelementptr inbounds i8, ptr %16, i64 48
  store ptr null, ptr %50, align 8, !tbaa.struct !40
  call void @_ZZZ10radix_sortRN7hipsycl4sycl5queueERNS0_6bufferIjLi1ESaIjEEEENKUlRNS0_7handlerEE0_clES8_ENKUlNS0_7nd_itemILi1EEEE_clESB_(ptr noundef nonnull align 8 dereferenceable(112) %17, ptr noundef nonnull byval(%"struct.hipsycl::sycl::nd_item") align 8 %16) #1
  call void @llvm.lifetime.end.p0(i64 8, ptr nonnull %15) #1
  call void @llvm.lifetime.end.p0(i64 64, ptr nonnull %16)
  call void @llvm.lifetime.end.p0(i64 112, ptr nonnull %17) #1
  ret void
}

; Function Attrs: nounwind
define void @_Z21__hipsycl_sscp_kernelIN7hipsycl4glue15__sscp_dispatch27ndrange_parallel_for_offsetIZZ10radix_sortRNS0_4sycl5queueERNS4_6bufferIjLi1ESaIjEEEENKUlRNS4_7handlerEE0_clESC_EUlNS4_7nd_itemILi1EEEE_Li1EEEEvRKT_(i64 %0, i64 %1, ptr %2, ptr %3, i64 %4, i64 %5, i64 %6, ptr %7, ptr %8, i64 %9, i32 %10, ptr %11, ptr %12, i64 %13, i64 %14) local_unnamed_addr #1 {
  %16 = alloca %"struct.hipsycl::sycl::nd_item", align 8
  %17 = alloca %"class.hipsycl::glue::__sscp_dispatch::ndrange_parallel_for_offset.776", align 8
  call void @llvm.lifetime.start.p0(i64 120, ptr nonnull %17) #1
  store i64 %0, ptr %17, align 8
  %18 = getelementptr inbounds i8, ptr %17, i64 8
  store i64 %1, ptr %18, align 8
  %19 = getelementptr inbounds i8, ptr %17, i64 16
  store ptr %2, ptr %19, align 8
  %20 = getelementptr inbounds i8, ptr %17, i64 24
  store ptr %3, ptr %20, align 8
  %21 = getelementptr inbounds i8, ptr %17, i64 32
  store i64 %4, ptr %21, align 8
  %22 = getelementptr inbounds i8, ptr %17, i64 40
  store i64 %5, ptr %22, align 8
  %23 = getelementptr inbounds i8, ptr %17, i64 48
  store i64 %6, ptr %23, align 8
  %24 = getelementptr inbounds i8, ptr %17, i64 56
  store ptr %7, ptr %24, align 8
  %25 = getelementptr inbounds i8, ptr %17, i64 64
  store ptr %8, ptr %25, align 8
  %26 = getelementptr inbounds i8, ptr %17, i64 72
  store i64 %9, ptr %26, align 8
  %27 = getelementptr inbounds i8, ptr %17, i64 80
  store i32 %10, ptr %27, align 8
  %28 = getelementptr inbounds i8, ptr %17, i64 88
  store ptr %11, ptr %28, align 8
  %29 = getelementptr inbounds i8, ptr %17, i64 96
  store ptr %12, ptr %29, align 8
  %30 = getelementptr inbounds i8, ptr %17, i64 104
  store i64 %13, ptr %30, align 8
  %31 = getelementptr inbounds i8, ptr %17, i64 112
  store i64 %14, ptr %31, align 8
  call void @llvm.lifetime.start.p0(i64 64, ptr nonnull %16)
  %32 = tail call i64 @__hipsycl_sscp_get_group_id_x() #1
  %33 = shl i64 %32, 32
  %34 = ashr exact i64 %33, 32
  %35 = tail call i64 @__hipsycl_sscp_get_local_id_x() #1
  %36 = shl i64 %35, 32
  %37 = ashr exact i64 %36, 32
  %38 = tail call i64 @__hipsycl_sscp_get_local_size_x() #1
  %39 = shl i64 %38, 32
  %40 = ashr exact i64 %39, 32
  %41 = tail call i64 @__hipsycl_sscp_get_num_groups_x() #1
  %42 = shl i64 %41, 32
  %43 = ashr exact i64 %42, 32
  %44 = mul nsw i64 %40, %34
  %45 = add nsw i64 %44, %37
  store ptr %31, ptr %16, align 8, !tbaa.struct !34
  %46 = getelementptr inbounds i8, ptr %16, i64 8
  store i64 %34, ptr %46, align 8, !tbaa.struct !35
  %47 = getelementptr inbounds i8, ptr %16, i64 16
  store i64 %37, ptr %47, align 8, !tbaa.struct !36
  %48 = getelementptr inbounds i8, ptr %16, i64 24
  store i64 %40, ptr %48, align 8, !tbaa.struct !37
  %49 = getelementptr inbounds i8, ptr %16, i64 32
  store i64 %43, ptr %49, align 8, !tbaa.struct !38
  %50 = getelementptr inbounds i8, ptr %16, i64 40
  store i64 %45, ptr %50, align 8, !tbaa.struct !39
  %51 = getelementptr inbounds i8, ptr %16, i64 48
  store ptr null, ptr %51, align 8, !tbaa.struct !40
  call void @_ZZZ10radix_sortRN7hipsycl4sycl5queueERNS0_6bufferIjLi1ESaIjEEEENKUlRNS0_7handlerEE0_clES8_ENKUlNS0_7nd_itemILi1EEEE_clESB_(ptr noundef nonnull align 8 dereferenceable(112) %17, ptr noundef nonnull byval(%"struct.hipsycl::sycl::nd_item") align 8 %16) #1
  call void @llvm.lifetime.end.p0(i64 64, ptr nonnull %16)
  call void @llvm.lifetime.end.p0(i64 120, ptr nonnull %17) #1
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind readnone speculatable willreturn
declare i32 @llvm.smax.i32(i32, i32) #4

attributes #0 = { argmemonly nocallback nofree nosync nounwind willreturn }
attributes #1 = { nounwind }
attributes #2 = { convergent nounwind }
attributes #3 = { inlinehint nounwind }
attributes #4 = { nocallback nofree nosync nounwind readnone speculatable willreturn }
attributes #5 = { mustprogress nounwind }
attributes #6 = { alwaysinline nounwind }

!llvm.module.flags = !{!0, !1, !2, !3}
!llvm.ident = !{!4}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 7, !"PIC Level", i32 2}
!2 = !{i32 7, !"PIE Level", i32 2}
!3 = !{i32 7, !"uwtable", i32 2}
!4 = !{!"clang version 15.0.7"}
!5 = !{!6, !7, i64 0}
!6 = !{!"_ZTSN7hipsycl4sycl8accessorIjLi1ELNS0_11access_modeE2ELNS0_6targetE3ELNS0_16accessor_variantE0EEE", !7, i64 0, !10, i64 8}
!7 = !{!"long", !8, i64 0}
!8 = !{!"omnipotent char", !9, i64 0}
!9 = !{!"Simple C++ TBAA"}
!10 = !{!"_ZTSN7hipsycl4sycl5rangeILi1EEE", !11, i64 0}
!11 = !{!"_ZTSN7hipsycl4sycl6detail12device_arrayImLm1EEE", !7, i64 0}
!12 = !{!13, !13, i64 0}
!13 = !{!"int", !8, i64 0}
!14 = !{!15, !15, i64 0}
!15 = !{!"any pointer", !8, i64 0}
!16 = !{!17, !13, i64 40}
!17 = !{!"_ZTSZZ10radix_sortRN7hipsycl4sycl5queueERNS0_6bufferIjLi1ESaIjEEEENKUlRNS0_7handlerEE_clES8_EUlNS0_7nd_itemILi1EEEE_", !6, i64 0, !18, i64 16, !13, i64 40, !22, i64 48}
!18 = !{!"_ZTSN7hipsycl4sycl8accessorIjLi1ELNS0_11access_modeE0ELNS0_6targetE0ELNS0_16accessor_variantE5EEE", !19, i64 0, !21, i64 16}
!19 = !{!"_ZTSN7hipsycl4sycl6detail13accessor_baseIjEE", !20, i64 0}
!20 = !{!"_ZTSN7hipsycl4glue16embedded_pointerIjEE", !8, i64 0}
!21 = !{!"_ZTSN7hipsycl4sycl6detail19conditional_storageINS1_8accessor38conditional_storage_buffer_range_tag_tELb1ENS0_5rangeILi1EEEEE", !10, i64 0}
!22 = !{!"_ZTSN7hipsycl4sycl8accessorIjLi1ELNS0_11access_modeE1ELNS0_6targetE0ELNS0_16accessor_variantE5EEE", !19, i64 0, !21, i64 16}
!23 = !{!7, !7, i64 0}
!24 = !{i64 0, i64 8, !23}
!25 = !{!26, !15, i64 0}
!26 = !{!"_ZTSN7hipsycl4sycl7nd_itemILi1EEE", !15, i64 0, !27, i64 8, !27, i64 16, !10, i64 24, !10, i64 32, !27, i64 40, !15, i64 48, !15, i64 56}
!27 = !{!"_ZTSN7hipsycl4sycl2idILi1EEE", !11, i64 0}
!28 = !{!29, !13, i64 80}
!29 = !{!"_ZTSZZ10radix_sortRN7hipsycl4sycl5queueERNS0_6bufferIjLi1ESaIjEEEENKUlRNS0_7handlerEE0_clES8_EUlNS0_7nd_itemILi1EEEE_", !6, i64 0, !18, i64 16, !6, i64 40, !18, i64 56, !13, i64 80, !22, i64 88}
!30 = !{i32 0, i32 33}
!31 = distinct !{!31, !32}
!32 = !{!"llvm.loop.mustprogress"}
!33 = !{!11, !7, i64 0}
!34 = !{i64 0, i64 8, !14, i64 8, i64 8, !23, i64 16, i64 8, !23, i64 24, i64 8, !23, i64 32, i64 8, !23, i64 40, i64 8, !23, i64 48, i64 8, !14, i64 56, i64 8, !14}
!35 = !{i64 0, i64 8, !23, i64 8, i64 8, !23, i64 16, i64 8, !23, i64 24, i64 8, !23, i64 32, i64 8, !23, i64 40, i64 8, !14, i64 48, i64 8, !14}
!36 = !{i64 0, i64 8, !23, i64 8, i64 8, !23, i64 16, i64 8, !23, i64 24, i64 8, !23, i64 32, i64 8, !14, i64 40, i64 8, !14}
!37 = !{i64 0, i64 8, !23, i64 8, i64 8, !23, i64 16, i64 8, !23, i64 24, i64 8, !14, i64 32, i64 8, !14}
!38 = !{i64 0, i64 8, !23, i64 8, i64 8, !23, i64 16, i64 8, !14, i64 24, i64 8, !14}
!39 = !{i64 0, i64 8, !23, i64 8, i64 8, !14, i64 16, i64 8, !14}
!40 = !{i64 0, i64 8, !14, i64 8, i64 8, !14}
```
Maybe we could dump exactly which line PTX compilation failed when raising such errors? Or at least provide the mangle kernel tag name when cuModuleLoadEx failed, not sure if there is a log query interface for this.

Reply:
Hi, thanks for the feedback. SYCL 2020 group algorithms like `inclusive_scan_over_group` are not yet implemented in the SSCP path. So you end up with calls to undefined functions, and then a JIT error (in the form of this CUDA error) is expected. So probably everything is working as expected :)